prepared_cache.go

```diff
@@ -1,37 +1,101 @@
 package gocql
 
 import (
-	"github.com/gocql/gocql/internal/lru"
 	"sync"
+	"sync/atomic"
+
+	"github.com/gocql/gocql/internal/lru"
 )
 
 const defaultMaxPreparedStmts = 1000
 
-// preparedLRU is the prepared statement cache
-type preparedLRU struct {
-	mu  sync.Mutex
-	lru *lru.Cache
+type preparedCache struct {
+	hosts     atomic.Value // map[string]*preparedLRU
+	mu        sync.Mutex
+	maxStmnts int
+}
+
+func (p *preparedCache) load() map[string]*preparedLRU {
+	v, _ := p.hosts.Load().(map[string]*preparedLRU)
+	return v
 }
 
-// Max adjusts the maximum size of the cache and cleans up the oldest records if
-// the new max is lower than the previous value. Not concurrency safe.
-func (p *preparedLRU) max(max int) {
+func (p *preparedCache) clear() {
 	p.mu.Lock()
-	defer p.mu.Unlock()
+	p.hosts.Store(make(map[string]*preparedLRU))
+	p.mu.Unlock()
+}
+
+func (p *preparedCache) forHost(addr string) *preparedLRU {
+	m := p.load()
+	lru, ok := m[addr]
+	if !ok {
+		p.mu.Lock()
+		m := p.load()
+		lru, ok = m[addr]
+		if !ok {
+			lru = p.addHost(m, addr)
+		}
+		p.mu.Unlock()
+	}
 
-	for p.lru.Len() > max {
-		p.lru.RemoveOldest()
+	return lru
+}
+
+func (p *preparedCache) addHost(current map[string]*preparedLRU, addr string) *preparedLRU {
+	cow := make(map[string]*preparedLRU, len(current)+1)
+	for k, v := range current {
+		cow[k] = v
 	}
-	p.lru.MaxEntries = max
+
+	lru := &preparedLRU{lru: lru.New(p.maxStmnts)}
+	cow[addr] = lru
+	p.hosts.Store(cow)
+	return lru
 }
 
-func (p *preparedLRU) clear() {
+func (p *preparedCache) HostUp(host *HostInfo) {
+	addr := JoinHostPort(host.ConnectAddress().String(), host.Port())
+
+	v := p.load()
+	if _, ok := v[addr]; ok {
+		return
+	}
+
+	cow := make(map[string]*preparedLRU, len(v)+1)
+	for k, v := range v {
+		cow[k] = v
+	}
+
+	cow[addr] = &preparedLRU{lru: lru.New(p.maxStmnts)}
+	p.hosts.Store(cow)
+}
+
+func (p *preparedCache) HostDown(host *HostInfo) {
+	addr := JoinHostPort(host.ConnectAddress().String(), host.Port())
+
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	for p.lru.Len() > 0 {
-		p.lru.RemoveOldest()
+	v := p.load()
+	if _, ok := v[addr]; !ok {
+		return
+	}
+
+	cow := make(map[string]*preparedLRU, len(v)-1)
+	for k, v := range v {
+		if k != addr {
+			cow[k] = v
+		}
 	}
+
+	p.hosts.Store(cow)
```

Comment:
There are many places where the map is modified. Perhaps a unified small path to add/remove where they all happen?